sdks/python/apache_beam/io/iobase.py

```diff
@@ -1402,20 +1402,36 @@ class _SDFBoundedSourceWrapper(ptransform.PTransform):
 
   NOTE: This transform can only be used with beam_fn_api enabled.
   """
+
+  class _SDFBoundedSourceRestriction(object):
+    """ A restriction wraps SourceBundle and RangeTracker. """
+    def __init__(self, source_bundle, range_tracker=None):
+      self.source_bundle = source_bundle
+      self.range_tracker = range_tracker
+
+    def __reduce__(self):
+      # The instance of RangeTracker shouldn't be serialized.
+      return (self.__class__, (self.source_bundle, ))
+
+
   class _SDFBoundedSourceRestrictionTracker(RestrictionTracker):
     """An `iobase.RestrictionTracker` implementations for wrapping BoundedSource
-    with SDF.
+    with SDF. The tracked restriction is a (SourceBundle, RangeTracker) pair.
+    In order to save bytes sent across the wire, the RangeTracker is set as
+    system tracking RangeTracker only when current_restriction is called.
 
     Delegated RangeTracker guarantees synchronization safety.
     """
     def __init__(self, restriction):
-      if not isinstance(restriction, SourceBundle):
+      if not isinstance(restriction,
+                        _SDFBoundedSourceWrapper._SDFBoundedSourceRestriction):
         raise ValueError('Initializing SDFBoundedSourceRestrictionTracker'
-                         'requires a SourceBundle')
-      self._delegate_range_tracker = restriction.source.get_range_tracker(
-          restriction.start_position, restriction.stop_position)
-      self._source = restriction.source
-      self._weight = restriction.weight
+                         ' requires a _SDFBoundedSourceRestriction')
+      self._source = restriction.source_bundle.source
+      self._weight = restriction.source_bundle.weight
+      self._delegate_range_tracker = self._source.get_range_tracker(
```

Comment:
How about just adding a range_tracker() method to _SDFBoundedSourceRestriction? Then _delegate_range_tracker would not have to be tracked here at all.